src/lib.rs

```diff
@@ -702,14 +703,28 @@ pub fn delegate(tokens: TokenStream) -> TokenStream {
     let block: DelegationBlock = syn::parse_macro_input!(tokens);
     let sections = block.segments.iter().map(|delegator| {
         let delegator_attribute = &delegator.delegator;
-
         let functions = delegator.methods.iter().map(|method| {
             let input = &method.method;
-            let signature = &input.sig;
-
+            let mut signature = input.sig.clone();
+            if let syn::Expr::Closure(closure) = delegator_attribute {
+                let additional_input: Vec<FnArg> = closure
+                    .inputs
+                    .iter()
+                    .map(|input| {
+                        if let syn::Pat::Type(pat_type) = input {
+                            syn::parse_quote!(#pat_type)
+                        } else {
+                            unreachable!()
+                        }
+                    })
+                    .collect();
+                let mut origin_inputs = mem::take(&mut signature.inputs).into_iter();
+                signature.inputs.push(origin_inputs.next().unwrap());
```

Comment:
You could also delegate associated methods, like this:
```rust
impl Wrapper {
    delegate! {
        to |k: u32| SomeStaticVariable {
            fn method() -> u32;
        }
    }
}
```
So the code shouldn't assume that `origin_inputs` is non-empty. It should check if the first argument is `self` (in any of its various forms), and only prepend it to the reconstructed input list if that is the case.

Reply:
Note that there are currently no tests for delegating associated methods I think, if you're up for it, it would be great to add some. Ideally also in combination with the closure.